### PR TITLE
8311084: Add typeSymbol() API for applicable constant pool entries

### PR DESCRIPTION
**Pull request by liach**
5 Constant Pool entries, namely ConstantDynamicEntry, InvokeDynamicEntry, FieldRefEntry, MethodRefEntry, and InterfaceMethodRefEntry should have a typeSymbol() API to return the nominal/symbolic descriptor (ClassDesc or MethodTypeDesc).

This API is not added to NameAndTypeEntry itself, for a NameAndTypeEntry only knows if its type should be a field or method type from the other entries that refer to it.

This is one of the issues discussed in this mailing list thread: https://mail.openjdk.org/pipermail/classfile-api-dev/2023-June/000381.html

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311084](https://bugs.openjdk.org/browse/JDK-8311084): Add typeSymbol() API for applicable constant pool entries (**Enhancement** - P4)

### Reviewers
 * [Brian Goetz](https://openjdk.org/census#briangoetz) (@briangoetz - **Reviewer**) ⚠️ Review applies to [d2701993](https://git.openjdk.org/jdk/pull/14706/files/d2701993bf19889f288f51a98f6ffd1a04fc025f)
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14706/head:pull/14706` \
`$ git checkout pull/14706`

Update a local copy of the PR: \
`$ git checkout pull/14706` \
`$ git pull https://git.openjdk.org/jdk.git pull/14706/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14706`

View PR using the GUI difftool: \
`$ git pr show -t 14706`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14706.diff">https://git.openjdk.org/jdk/pull/14706.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14706#issuecomment-1613165476)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/classfile/constantpool/ConstantDynamicEntry.java [hunk 1]
```diff
@@ -39,6 +41,13 @@ public sealed interface ConstantDynamicEntry
         extends DynamicConstantPoolEntry, LoadableConstantEntry
         permits AbstractPoolEntry.ConstantDynamicEntryImpl {
 
+    /**
+     * {@return a symbolic descriptor for the dynamic constant's type}
+     */
+    default ClassDesc typeSymbol() {
+        return Util.fieldTypeSymbol(nameAndType());
+    }
```
Comment (briangoetz) [inline src/java.base/share/classes/jdk/internal/classfile/constantpool/ConstantDynamicEntry.java:49]: Usage of Util::fieldTypeSymbol in asSymbol() can be replaced with typeSymbol()

File: src/java.base/share/classes/jdk/internal/classfile/constantpool/InvokeDynamicEntry.java [hunk 2]
```diff
@@ -37,6 +38,13 @@ public sealed interface InvokeDynamicEntry
         extends DynamicConstantPoolEntry
         permits AbstractPoolEntry.InvokeDynamicEntryImpl {
 
+    /**
+     * {@return a symbolic descriptor for the call site's invocation type}
+     */
+    default MethodTypeDesc typeSymbol() {
+        return Util.methodTypeSymbol(nameAndType());
+    }
```
Comment (briangoetz) [inline src/java.base/share/classes/jdk/internal/classfile/constantpool/InvokeDynamicEntry.java:46]: Similar comment as with CondyEntry -- reduce usage of Util::methodTypeSymbol in this class

Comment (briangoetz) [reply src/java.base/share/classes/jdk/internal/classfile/constantpool/InvokeDynamicEntry.java:46]: (Would like to do the same in various ::count methods, but unfortunately MemberRefEntry is too general; would have to rejigger the various XxxRefEntry types.)

**Review by briangoetz — APPROVED** [review]

File: src/java.base/share/classes/jdk/internal/classfile/constantpool/InvokeDynamicEntry.java (hunk 2 above)
Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/constantpool/InvokeDynamicEntry.java:46]: Which `count` methods are you referring to, Brian?

**Comment by liach** [comment]
@asotona Can you review this Classfile API utility addition?

**Review by asotona — APPROVED** [review]
It looks good, thanks for the patch.

**Pull request closed without merging**